R/gibbs_draws.R

```diff
@@ -1,3 +1,99 @@
+#'   (stored in element (l-1)*(t_max - t_min + 1) + t of the vector)
+#'
+#' @export
+#'
+#' @examples
+#'
+#' n_v <- 2
+#' n_loc <- 3 # 3 locations
+#' T <- 100 # 100 time steps
+#' priors <- default_priors()
+#' # constant incidence 10 per day everywhere
+#' incid <- array(10, dim = c(T, n_loc, n_v))
+#' get_shape_R_flat(incid, priors)
+#'
+get_shape_R_flat <- function(incid, priors, t_min = 2L, t_max = nrow(incid)) {
+  t <- seq(t_min, t_max, 1)
+  shape <- apply(incid[t, , , drop = FALSE], c(1, 2), sum) + priors$R$shape 
+  as.numeric(shape)
+}
+
+#' Precompute shape of posterior distribution for epsilon
+#'
+#' @param incid a multidimensional array containing values of the (local)
+#'   incidence
+#'   for each time step (1st dimension), location (2nd dimension) and
+#'   pathogen/strain/variant (3rd dimension)
+#'
+#' @param lambda a multidimensional array containing values of the overall
+#'   infectivity for each time step (1st dimension), location (2nd dimension)
+#'   and pathogen/strain/variant (3rd dimension). The overall infectivity for
+#'   a given location and pathogen/strain/variant represents the sum of
+#'   the incidence for that location and that pathogen/strain/variant at all
+#'   previous time steps, weighted by the current infectivity of those
+#'   past incident cases. It can be calculated from the incidence `incid` and
+#'   the distribution of the serial interval using function `compute_lambda`)
+#'
+#' @param priors a list of prior parameters (shape and scale of a gamma
+#'   distribution) for epsilon and R; can be obtained from the function
+#'   `default_priors`. The prior for R is assumed to be the same for all
+#'   time steps and all locations
+#'
+#' @param t_min an integer >1 giving the minimum time step to consider in the
+#'   estimation. Default value is 2 (as the estimation is conditional on
+#'   observations at time step 1 and can therefore only start at time step 2).
+#'
+#' @param t_max an integer >`t_min` and <=`nrow(incid)` giving the maximum time
+#'   step to consider in the estimation. Default value is `nrow(incid)`.
+#'
+#' @return a value or vector of values of the shape of the posterior 
+#'   distribution of epsilon for each of the non reference variants
+#'
+#' @export
+#'
+#' @examples
+#'
+#' n_loc <- 4 # 4 locations
+#' n_v <- 3 # 3 strains
+#' T <- 100 # 100 time steps
+#' priors <- default_priors()
+#' # constant incidence 10 per day everywhere
+#' incid <- array(10, dim = c(T, n_loc, n_v))
+#' incid <- process_I_multivariant(incid)
+#' # arbitrary serial interval, same for both variants
+#' w_v <- c(0, 0.2, 0.5, 0.3)
+#' si_distr <- cbind(w_v, w_v, w_v)
+#' lambda <- compute_lambda(incid, si_distr)
+#' get_shape_epsilon(incid$local, lambda, priors)
+#'
+get_shape_epsilon <- function(incid, lambda, priors,
+                              t_min = 2L, t_max = nrow(incid)) {
+  t <- seq(t_min, t_max, 1)
+  EpiEstim:::vnapply(seq(2, dim(lambda)[3]), function(e)
```

Comment:
I get warnings on running devtools::check about the use of ::: operator. This might be an issue for submission to CRAN. Warning reads:
 There are ::: calls to the package's namespace in its code. A package
    almost never needs to use ::: for its own objects:
    ‘vnapply’

Reply:
ok; but this was already in the code I just moved it! So I think the issue predates my PR. I will remove